tensorflow/core/common_runtime/pluggable_device/pluggable_device.cc

```diff
@@ -0,0 +1,425 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <stdlib.h>
+#include <string.h>
+#include <algorithm>
+#include <list>
+#include <map>
+#include <tuple>
+#include <vector>
+
+#include "tensorflow/core/common_runtime/device/device_event_mgr.h"
+#include "tensorflow/core/common_runtime/device/device_id.h"
+#include "tensorflow/core/common_runtime/device/device_id_manager.h"
+#include "tensorflow/core/common_runtime/device/device_id_utils.h"
+#include "tensorflow/core/common_runtime/device_factory.h"
+#include "tensorflow/core/common_runtime/local_device.h"
+#include "tensorflow/core/common_runtime/pluggable_device/pluggable_device.h"
+#include "tensorflow/core/common_runtime/pluggable_device/pluggable_device_context.h"
+#include "tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.h"
+#include "tensorflow/core/common_runtime/pluggable_device/pluggable_device_init.h"
+#include "tensorflow/core/common_runtime/pluggable_device/pluggable_device_process_state.h"
+#include "tensorflow/core/common_runtime/pluggable_device/pluggable_device_util.h"
+#include "tensorflow/core/framework/allocator.h"
+#include "tensorflow/core/framework/device_base.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/tensor.h"
+#include "tensorflow/core/framework/tensor.pb.h"
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/framework/variant_op_registry.h"
+#include "tensorflow/core/graph/types.h"
+#include "tensorflow/core/lib/core/errors.h"
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/lib/strings/numbers.h"
+#include "tensorflow/core/lib/strings/str_util.h"
+#include "tensorflow/core/lib/strings/strcat.h"
+#include "tensorflow/core/platform/logging.h"
+#include "tensorflow/core/platform/macros.h"
+#include "tensorflow/core/platform/stream_executor.h"
+#include "tensorflow/core/platform/types.h"
+#include "tensorflow/core/public/session_options.h"
+#include "tensorflow/core/util/device_name_utils.h"
+#include "tensorflow/core/util/env_var.h"
+#include "tensorflow/core/util/stream_executor_util.h"
+
+namespace tensorflow {
+
+// This factory helps to ensure that different PluggableDevice objects that
+// refer to the same physical device and stream group id use the same stream
+// group object (and therefore the same device streams). This is necessary since
+// there is a single memory allocator per device (see
+// ProcessState::GetPluggableDeviceAllocator) and allocators must not be shared
+// across streams.
+class PluggableDevice::StreamGroupFactory {
```

Comment:
@sanjoy `PluggableDevice::StreamGroupFactory` is pretty similar to `BaseGPUDevice::StreamGroupFactory`. Should it be refactored to the `common_runtime/device` folder?